### DISCUSSION TITLE
Maybe a very novice question: how can I optimize Slint's performance and memory usage?

### DISCUSSION THREAD

**Discussion by Horbin-Magician** [slint-ui/slint#3376] [Q&A]
(I have so many questions. I don't know if it's right to raise them here. If it bothers you, please point out for me)

Environment: Windows + Rust + Slint (renderer-winit-skia)
A blank window will use about 50M of memory. This number increases dramatically as the number of windows increases.

I know that different backends will behave differently. I've tried every other backend except qt. No backend wins overall.
As a result, I couldn't get a program that was comparable to Qt programes in terms of file size, window startup speed, and memory usage.

I don't know if that's normal. In fact, I'm willing to sacrifice performance for fast development.
What I want to know is that if there is still room for optimization, and where I can start to optimize my program?

If you'd like, you can check out my project code https://github.com/Horbin-Magician/rotor-rs. This is a very rough project and will only work successfully on Windows system.

(Please forgive me if I have a lot of information left out. If there is any information that is not provided, please feel free to ask.)

**Comment by tronical** [comment]
Thanks for starting this discussion :). This is a good topic and I'm interesting in exchanging experience and measurements. Maybe they'll lead us to some discovery and perhaps opportunity of optimisation.

I don't quite have any data on start up time, but I have two data points on file sizes and memory consumption.

In terms of file sizes, this is what I see:

For me, on Windows, Qt6Core.dll is 5.6 MB, and if you add Gui, a Qt application will have a disk footprint starting at 13-14MB.
Our Slint Rust template starts for me at around 3.6 MB and with LTO, stripping, etc. I get it down to 2.6MB (although that's with the femtovg renderer, not Skia). I think that's pretty competitive as a starting point, since that's statically linked (no extra DLLs apart from Windows system DLLs needed).

In terms of memory consumption, one thing I notice is that the memory consumption appears to correlate with the size of the window. For men an _empty_ qmlscene window weighs in at around 35MB in the task manager. It starts out with less, but as I resize it to around 1200x800 that's where it ends up.

A similar minimal `export component App inherits Window {}` run with `slint-viewer.exe` and with `SLINT_BACKEND=Skia` clocks in at around 20MB at the same window size. It starts out with less memory though. With `SLINT_BACKEND=Qt` I get a similar consumption at around 18MB.

This is in a VMware virtual machine with Vmware graphics drivers.

My suspicion is that the memory consumption that you see is likely related to how your graphics drivers allocates window surface buffers. With Skia I think the D3D swap chain will have at least two buffers the size of the window.

Do you also observe a correlation between the memory consumption and the window size? What's your graphics driver setup?

**Reply by Horbin-Magician** [reply]
I'm sorry I've been a bit busy lately and can only respond now.
R1: Yes,  I also observed a correlation between the memory consumption and the window size. But different render behaves differently.
R2: I have test the programe in two machines. One use Intel(R) HD Graphics 530, the other use AMD Radeon 8 Graphics (Renoir) + Nvidia 3060. I have not changed any of the driver settings, and the two machines performed similarly.

Here is a detailed data table: (it may be useful)

Renderer--------Memory usage----File size
femtovg---------51.1MB------------3.45MB
skia-opengl-----52.6MB------------19.5MB
skia-d3d--------17.5MB-------------20.3MB
winit-software--14.0MB-------------3.87MB

> a. Test machines: 1080p displayer + Intel(R) HD Graphics 530 + Windows10 + Rust + Slint + slint-rust-template
> b. The memory usage of "femtovg" and "skia-opengl" will change dramatically when resize the window (easily increase to more than 200MB)
> c. I can't build my program when use winit-skia-vulkan, because cargo can't download the package "https://github.com/rust-skia/skia-binaries/releases/download/0.63.0/skia-binaries-e9c8bd6ba59d6d360bd9-x86_64-pc-windows-msvc-d3d-textlayout-vulkan.tar.gz".

It seems that my results are not too different from yours. There are just differences between different renderers.
skia-d3d is good for me, but it doesn't  support transparent Windows (it's true?).

**Comment by nic-obert** [comment]
Why do GUI applications take up so much memory? And what does memory consumption in a GUI app depend on? I'm quite new to GUI app development

**Reply by tronical** [reply]
In practice the biggest consumer of memory are rendering buffers that correlate with the window size. Depending on where they are stored and how memory consumption is measured, those can skew results because they tend to be orders of magnitude bigger than the rest of the UI related data structures typically allocated.
